Refactor dadadj for CCPP framework

jtruesdal opened the pull request:
Add dadadj ccpp routine meta data and suite xml files to atmos_phys.

nusbaume commented on suite_dadadj.xml:
```diff
@@ -0,0 +1,10 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Given that `dadadj` is the first true CAM7 physics scheme, I might vote to make a copy of this SDF called `suite_cam7.xml`.  I then might move the `suite_dadadj.xml` file to a new `test/test_sdfs` directory, as we will likely only use this particular SDF for testing purposes.

jtruesdal replied:
Temporarily left a suite_dadadj.xml in the original spot for testing but will remove before commit.  Updated the rest as recommended.

nusbaume reviewed (changes requested):
Great work @jtruesdal !  I do have some change requests, but  please let me know if you have any questions or concerns with any of them.  Thanks!

jtruesdal commented on dadadj/dadadj.F90:
```diff
@@ -0,0 +1,199 @@
+    errflg = 0
+    scheme_name = 'DADADJ'
+
+    allocate(c1dad(nlvdry), c2dad(nlvdry), c3dad(nlvdry), c4dad(nlvdry))
```
Added all the stat checks and endruns.

jtruesdal commented on dadadj/dadadj.F90:
```diff
@@ -0,0 +1,199 @@
+       cappaint = 0.5_kind_phys*(cappa(i,2) + cappa(i,1))
+       gammad = cappaint*0.5_kind_phys*(t(i,2) + t(i,1))/pint(i,2)
+       dtdp = (t(i,2) - t(i,1))/(pmid(i,2) - pmid(i,1))
+       dodad(i) = (dtdp + zeps) .gt. gammad
```
Done.

jtruesdal commented on dadadj/dadadj.F90:
```diff
@@ -0,0 +1,199 @@
+            zeps = zeps + zeps
+            if (zeps > 1.e-4_kind_phys) then
+               errflg = i
+               return                ! error return
```
Yes. Done.

jtruesdal commented on dadadj/dadadj.meta:
```diff
@@ -0,0 +1,134 @@
+[ errflg ]
+  standard_name = ccpp_error_flag
+  long_name = Error flag for error handling in CCPP
+  units = flag
```
Cone.

jtruesdal replied:
Done

jtruesdal commented on dadadj/dadadj.meta:
```diff
@@ -0,0 +1,134 @@
+  dimensions = (horizontal_dimension, vertical_layer_dimension)
+  intent = in
+[ state_q ]
+  standard_name = water_vapor_mixing_ratio_wrt_dry_air
```
Done.

jtruesdal commented on dadadj/dadadj.meta:
```diff
@@ -0,0 +1,134 @@
+  intent = out
+[ dadpdf ]
+  standard_name = indicator_for_dry_adiabatic_adjusted_grid_cell
+  units = 1
```
Done

jtruesdal commented on dadadj/dadadj.meta:
```diff
@@ -0,0 +1,134 @@
+  standard_name = air_temperature
+  type = real | kind = kind_phys
+  units = K
+  dimensions = (horizontal_dimension, vertical_layer_dimension)
```
Renamed all the instances of horizontal_dimension to horizontal_loop_extent

jtruesdal commented on dadadj/dadadj.meta:
```diff
@@ -0,0 +1,134 @@
+[ errflg ]
+  standard_name = ccpp_error_flag
+  long_name = Error flag for error handling in CCPP
+  units = flag
```
Done.

jtruesdal commented on dadadj/dadadj_namelist.xml:
```diff
@@ -0,0 +1,108 @@
+  <standard_name>number_of_vertical_levels_from_model_top_where_dry_adiabatic_adjustment_occurs</standard_name>
+  <units>count</units>
+  <desc>
+    Number of layers from the top of the model over which to do dry convective adjustment.
```
Done.

jtruesdal commented on dadadj/dadadj_namelist.xml:
```diff
@@ -0,0 +1,108 @@
+  <units>count</units>
+  <desc>
+    The maximum number of iterations to achieve convergence in dry adiabatic adjustment.
+    For WACCM-X it can be advantageous to use a number which is much higher than the CAM
```
Done

jtruesdal commented on doc/ChangeLog:
```diff
@@ -1,6 +1,52 @@
 
 ===============================================================
 
+Tag name: atmos_phys0_02_xxx
```
Done.

jtruesdal commented on doc/ChangeLog:
```diff
@@ -1,6 +1,52 @@
+A	suite_dadadj.xml - CCPP suite file
+
+
+List and Describe any test failures:
+    - Tested in CAM
+	Because dadadj doesn't normally get tripped I modified the Temp
+	profile to create an instability where the code would be exercized.
+	The snapshot files continain this instability.  The modification adds
+	60 degrees to the layer 2 temperature.
+
+	--- /project/amp/jet/collections/cam6_3_160_dryadj.042924.1856/src/dynamics/se/dyn_comp.F90	2024-05-02 16:56:49.746341064 -0600
+	+++ SourceMods/src.cam/dyn_comp.F90	2024-05-02 16:37:13.171433366 -0600
+	@@ -1429,6 +1429,7 @@
+          do j = 1, np
+             do i = 1, np
+                elem(ie)%state%T(i,j,:,1) = dbuf3(indx,:,ie)
++               elem(ie)%state%T(i,j,2,1) = dbuf3(indx,2,ie) + 60._r8
+                indx = indx + 1
+             end do
+          end do
```
Created a dry_adiabatic_adjust directory and put the modified file underneath.

nusbaume commented on dry_adiabatic_adjust/dadadj.F90:
```diff
@@ -38,7 +38,8 @@ subroutine dadadj_init(dadadj_nlvdry, dadadj_niter, nz, errmsg, errflg)
        errflg = 1
-       write(errmsg,*) 'dadadj_init: dadadj_nlvdry=',dadadj_nlvdry,' but must be less than the number of vertical levels '
+       write(errmsg,*) 'dadadj_init: dadadj_nlvdry=',dadadj_nlvdry,' but must be less than the number of vertical levels ',&
+       '(',nz,'), and must be a positive integer.`
```
Whoops, I think I had a typo for this particular suggestion:

```suggestion
       '(',nz,'), and must be a positive integer.'
```

jtruesdal replied:
Done.

nusbaume commented on suite_cam7.xml:
```diff
@@ -0,0 +1,10 @@
+
+<suite name="dadadj" version="1.0">
+  <group name="physics_before_coupler">
+    <scheme>dadadj</scheme>
```
I might add a comment here stating that this is dry adiabatic adjustment:

```suggestion
    <!-- Dry Adiabatic Adjustment -->
    <scheme>dadadj</scheme>
```

jtruesdal replied:
Done.

nusbaume commented on dry_adiabatic_adjust/dadadj.meta:
```diff
@@ -126,9 +126,9 @@
-  units = flag
+  standard_name = ccpp_error_code
+  long_name = Error code for error handling in CCPP
+  units = 1
```
Just adding a second reminder to add an updated version of the `NamesNotInDictionary.txt` file to this PR following the instructions found online here:

https://github.com/ESCOMP/atmospheric_physics/wiki/Development-workflow#namesnotindictionarytxt

jtruesdal replied:
Done.

nusbaume reviewed (changes requested):
Super-close!  I just have a few last (hopefully very minor) requests.

nusbaume commented on suite_cam7.xml:
```diff
@@ -0,0 +1,10 @@
+<?xml version="1.0" encoding="UTF-8"?>
+
+<suite name="dadadj" version="1.0">
```
Whoops, forgot to mention in my previous review that this should probably be changed to `cam7` as well:

```suggestion
<suite name="cam7" version="1.0">
```

jtruesdal commented on dadadj/dadadj.meta:
```diff
@@ -0,0 +1,134 @@
+  units = flag
+  type = integer
+  dimensions = ()
+  intent = out
```
Thanks updated.

nusbaume reviewed (approved):
Everything looks great now, thanks!

cacraigucar commented on dry_adiabatic_adjust/dadadj.F90:
```diff
@@ -0,0 +1,219 @@
+         zeps = 2.0e-5_kind_phys
+
+         do k = 1, nlvdry
+            cappaint = 0.5_kind_phys*(cappa(i,k+1) + cappa(i,k))
```
Was the addition of this line a bug that needed to be fixed? Does it not change answers between this and what is currently in CAM?

jtruesdal replied:
I think it is a bug, introduced when variable cappa was added to the model.  I've sent this off to Peter to see if he agrees and if he needs to see a control before this is approved.

jtruesdal replied:
Peter approved the change and doesn't think we need a control.

jtruesdal commented on doc/ChangeLog:
```diff
@@ -1,6 +1,52 @@
+A	dadadj/dadadj.meta
+A	dadadj/dadadj_namelist.xml
+M	doc/ChangeLog
+A	suite_dadadj.xml - CCPP suite file
```
changed the directory names to dry_adiabatic_adjust.  Is that what you meant?

cacraigucar replied:
No - the suite name changed and needs to be changed here as well

jtruesdal replied:
Done.

cacraigucar replied:
Do you need to push the ChangeLog back?

jtruesdal replied:
I think this is OK.  When I first looked at it via this page it looked old but I refreshed and clicked on the file link via the lefthand frame and the names look to be spelled out.  I'm not sure why it wasn't updated the first time.

cacraigucar replied:
The line here says the file is suite_dryadj.xml.  The file that I see was added is suite_dry_adiabatic_adjust.xml.  I looked directly in your doc/ChangeLog to see what it says.  I think the filename listed in the ChangeLog needs to be updated.

jtruesdal replied:
Also the suite_dry_adiabatic_adjust.xml routine was moved to test/test_sdfs as requested by @nusbaume.  The suite_cam7.xml now contains dry_adiabatic_adjust as the first parameterization of the cam7 suite.  I think all future cam7 parameterizations will go into that suite as well as they are finished.  I'm happy to move things aound though if the plan has changed.

cacraigucar replied:
I agree.  This is just a reminder that this section of the ChangeLog needs to be updated to reflect the current location and names of the files.

jtruesdal replied:
Thanks for being patient. Updated Changelog with description and location of all 9 files, merged up to latest tag and regenerated NamesNotInDictionary.txt.

jtruesdal commented on doc/ChangeLog:
```diff
@@ -1,6 +1,52 @@
+List all existing files that have been added (A), modified (M), or deleted (D),
+and describe the changes:
+
+A	dadadj/dadadj.F90  - minor refactor to the cam routine - ccpp'ize
+A	dadadj/dadadj.meta
+A	dadadj/dadadj_namelist.xml
```
Changed the directory names to dry_adiabatic_adjust

jtruesdal commented on doc/ChangeLog:
```diff
@@ -1,6 +1,52 @@
+List and Describe any test failures:
+    - Tested in CAM
+	Because dadadj doesn't normally get tripped I modified the Temp
+	profile to create an instability where the code would be exercized.
```
corrected

jtruesdal commented on doc/ChangeLog:
```diff
@@ -1,6 +1,52 @@
+    - Tested in CAM
+	Because dadadj doesn't normally get tripped I modified the Temp
+	profile to create an instability where the code would be exercized.
+	The snapshot files continain this instability.  The modification adds
```
corrected

jtruesdal commented:
I've added a new temporary update routine to update the q tendency. 
 This will be removed once a standard utility for updating the constituents is in place.  I also modified the parameterization run method to return a heating tendency which now takes care of a roundoff issue in temperature.  Everything is again bfb.

jtruesdal commented:
@cacraigucar @nusbaume I added a new temporary run method to update the constituent array (dadadj_apply_qv_tendency.F90/meta).  Everything is working so hopefully this is the final review.  I'm placing these snapshot files in their proper location.

nusbaume commented on doc/ChangeLog:
```diff
@@ -1,5 +1,55 @@
 ===============================================================
 
+Tag name: atmos_phys0_0x_xxx
```
Just FYI that the tag will likely be `atmos_phys0_04_000`.

jtruesdal replied:
Thx. Updated.

nusbaume commented on doc/ChangeLog:
```diff
@@ -1,5 +1,55 @@
+Originator(s): jet
+Date: May 3, 2024
+One-line Summary: dadadj CCPP mods
+Github PR URL: hhttps://github.com/ESCOMP/atmospheric_physics/pull/
```
Just a reminder to update this URL:

```suggestion
Github PR URL: https://github.com/ESCOMP/atmospheric_physics/pull/91
```

jtruesdal replied:
Updated and added reviewers.

nusbaume commented on dry_adiabatic_adjust/dadadj.F90:
```diff
@@ -50,22 +50,24 @@ end subroutine dadadj_init
     real(kind_phys), intent(in) :: state_q(:,:)   ! specific humidity
-    real(kind_phys), intent(out), target :: t_tend(:,:)   ! temperature tendency
-    real(kind_phys), intent(out), target :: q_tend(:,:)   ! specific humidity tendency
+    real(kind_phys), intent(out) :: s_tend(:,:)   ! temperature tendency
```
I would change this comment to state that the tendency is for dry air enthalpy:

```suggestion
    real(kind_phys), intent(out) :: s_tend(:,:)   ! dry air enthalpy tendency
```

jtruesdal replied:
Fixed comment.

nusbaume commented on dry_adiabatic_adjust/dadadj_apply_qv_tendency.meta:
```diff
@@ -0,0 +1,42 @@
+  intent = in
+[ state_q ]
+  standard_name = water_vapor_mixing_ratio_wrt_moist_air_and_condensed_water
+  long_name = mass mixing ratio of water vapor / dry air
```
I think this long name needs to be updated:

```suggestion
  long_name = mass mixing ratio of water vapor / moist air and condensed water
```

jtruesdal replied:
Updated long_name

nusbaume reviewed (approved):
Looks good!  I just some minor requests to update a comment, a long name, and the ChangeLog (doesn't need a re-review though).
